File: typescript/base-hmi-types/src/screens/base/HmiPaintedScreenItemBase.ts
import { HmiProperty } from "./HmiProperty.js";
import { HmiColor } from "./HmiColor.js";
import { HmiScreenItemBase } from "./HmiScreenItemBase.js";
import { HmiThickness } from "./HmiThickness.js";
import { HmiFillAnimation } from "./HmiFillAnimation.js";
import { HmiColorAnimation } from "./HmiColorAnimation.js";

export abstract class HmiPaintedScreenItemBase extends HmiScreenItemBase {
  foregroundColor?: HmiProperty<HmiColor>;
  alternateForegroundColor?: HmiProperty<HmiColor>;
  backgroundColor?: HmiProperty<HmiColor>;
  alternateBackgroundColor?: HmiProperty<HmiColor>;
  borderColor?: HmiProperty<HmiColor>;
  borderBackgroundColor?: HmiProperty<HmiColor>;
  borderUsesBackgroundColor?: HmiProperty<boolean>;
  alternateBorderColor?: HmiProperty<HmiColor>;
  borderWidth?: HmiProperty<number>;
  borderStyle?: HmiProperty<number>;
  focusHighlight?: HmiProperty<boolean>;
  pointerHighlight?: HmiProperty<boolean>;
  backgroundBlink?: HmiProperty<boolean>;
  fillAnimation?: HmiFillAnimation;
  colorAnimation?: HmiColorAnimation;
  margin?: HmiThickness;
  padding?: HmiThickness | HmiProperty<number>;
}

File: src/BaseHmiTypes/Screens/Base/HmiPaintedScreenItemBase.cs
namespace BaseHmiTypes.Screens.Base;

public abstract class HmiPaintedScreenItemBase : HmiScreenItemBase
{
    public HmiProperty<HmiColor>? ForegroundColor { get; set; }

    public HmiProperty<HmiColor>? AlternateForegroundColor { get; set; }

    public HmiProperty<HmiColor>? BackgroundColor { get; set; }

    public HmiProperty<HmiColor>? AlternateBackgroundColor { get; set; }

    public HmiProperty<HmiColor>? BorderColor { get; set; }

    public HmiProperty<HmiColor>? BorderBackgroundColor { get; set; }

    /// <summary>
    /// Gets or sets whether the background color is also used for the object's border.
    /// </summary>
    public HmiProperty<bool>? BorderUsesBackgroundColor { get; set; }

    public HmiProperty<HmiColor>? AlternateBorderColor { get; set; }

    public HmiProperty<double>? BorderWidth { get; set; }

    public HmiProperty<int>? BorderStyle { get; set; }

    public HmiProperty<bool>? FocusHighlight { get; set; }

    public HmiProperty<bool>? PointerHighlight { get; set; }

    public HmiProperty<bool>? BackgroundBlink { get; set; }

    public HmiFillAnimation? FillAnimation { get; set; }

    public HmiColorAnimation? ColorAnimation { get; set; }

    public HmiThickness? Margin { get; set; }

    public HmiThickness? Padding { get; set; }
}
